accepted/single-file/design_3_0.md

```diff
@@ -0,0 +1,136 @@
+# Single-file Publish
+
+The design of single-file apps in .net core 3.0.
+
+## Introduction
+
+In .net core 3 implements Stage 1 (described in the [staging document](staging.md)) support for single-file apps.
+
+## Build System Interface
+
+Publishing to a single file can be triggered by adding the following property to an application's project file:
+
+```xml
+<PropertyGroup>
+    <PublishSingleFile>true</PublishSingleFile>
+</PropertyGroup>    
+```
+
+* The `PublishSingleFile` property applies to both framework dependent and self-contained publish operations.
+* The `PublishSingleFile` property applies to platform-specific builds with respect to a given runtime-identifier. The output of the build is a native binary for the specified platform.
+  When `PublishSingleFile`  is set to `true`, it is an error to leave `RuntimeIdentifier` undefined, or to set `UseAppHost` to `false`.
+* Setting the `PublishSingleFile`property causes the managed app, managed dependencies, platform-specific native dependencies, configurations, etc. (basically the contents of the publish directory when `dotnet publish` is run without setting the property) to be embedded within the native `apphost`. 
+
+By default, the symbol files are not embedded within the single-file, but remain as separate files in the publish directory. This includes both the IL `.pdb` file, and the native `.ni.pdb` / `app.guid.map` files generated by ready-to-run compiler. Setting the following property causes the symbol files to be included in the single-file.
+
+```xml
+<PropertyGroup>
+    <IncludeSymbolsInSingleFile>true</IncludeSymbolsInSingleFile>
```

Comment:
How does the diagnostic stack gets access to these embedded symbol files? Are they going to be extracted to temp during every startup? Or are all places going to be enlighten to read these from the single-file (this is hard in particular for the .ni.pdb and app.guid.map files - we do not have direct control over where these are read)

Reply:
Yes `IncludeSymbolsInSingleFile` is expected to be supported through extraction. 
In fact, for 3.0, all bundled files are extracted at startup (unless they are reused).

`IncludeSymbolsInSingleFile` will trigger extraction even in .net 5 version.

Reply:
Ok. Still, I think it is likely that there will be fixes in the diagnostic stack required to make the extracted files used by the diagnostic tools.

Reply:
I agree, In 3.0, the DLL and the PDB lived in the same extraction directory, so the symbols were loaded from the same directory. 

In 5.0, when the assemblies are loaded from memory, we might need to set the additional paths to locate the extracted symbol files. @mikem8361